ang/civicase/PanelQuery.js

```diff
@@ -0,0 +1,196 @@
+(function (angular, _) {
+  var module = angular.module('civicase');
+
+  module.directive('civicasePanelQuery', function () {
+    return {
+      restrict: 'E',
+      templateUrl: '~/civicase/PanelQuery.html',
+      link: linkFn,
+      controller: 'panelQueryCtrl',
+      scope: {
+        query: '<',
+        customData: '<?',
+        handlers: '<?'
+      },
+      transclude: {
+        actions: '?panelQueryActions',
+        results: 'panelQueryResults'
+      }
+    };
+
+    function linkFn ($scope, $element, $attrs, $controller, $transclude) {
+      $transclude($scope, function (clone, scope) {
+        $element.find('[ng-transclude="actions"]').html(clone);
+      }, false, 'actions');
+
+      $transclude($scope, function (clone, scope) {
+        $element.find('[ng-transclude="results"]').html(clone);
+      }, false, 'results');
+    }
+  });
+
+  module.controller('panelQueryCtrl', panelQueryCtrl);
+
+  panelQueryCtrl.$inject = ['$log', '$q', '$scope', 'crmApi'];
+
+  function panelQueryCtrl ($log, $q, $scope, crmApi) {
+    var PAGE_SIZE = 5;
+
+    $scope.customData = $scope.customData || {};
+    $scope.handlers = $scope.handlers || {};
+    $scope.loading = false;
+    $scope.results = [];
+    $scope.title = '---';
+    $scope.total = 0;
+    $scope.selectedRange = 'week';
+    $scope.periodRange = [
+      { label: 'This Week', value: 'week' },
+      { label: 'This Month', value: 'month' }
+    ];
+    $scope.pagination = {
+      page: 1,
+      size: PAGE_SIZE,
+      range: { from: 1, to: PAGE_SIZE }
+    };
+
+    (function init () {
+      initWatchers();
+      verifyData();
+
+      loadData();
+    }());
+
+    /**
+     * Calculates the offset of the results list based
+     * on the current page number and page size
+     *
+     * @return {String}
+     */
+    function calculatePageOffset () {
+      return ($scope.pagination.page - 1) * $scope.pagination.size;
+    }
+
+    /**
+     * Fetches the data via the Civi API and stores the response
+     *
+     * @param {Boolean} skipCount if true then the "getcount" request won't be sent
+     * @return {Promise}
+     */
+    function fetchDataViaApi (skipCount) {
+      var apiCalls = {
+        get: [ $scope.query.entity, $scope.query.action || 'get', prepareRequestParams() ],
```

Comment:
In my opinion it would make for a worse signature, by spreading all the different bits related to the query across different attributes. I think it's best to keep them grouped together